nemo/collections/asr/parts/preprocessing/perturb.py

```diff
@@ -583,6 +570,204 @@ def perturb_with_foreground_noise(self, data, noise, data_rms=None, max_noise_du
+
+        self.perturb_with_input_noise(data, noise, ref_mic=ref_mic, norm_to_db=self._norm_to_db)
+
+    def snr_mixer(self, clean, noise, snr, norm_to_db=-25.0):
+        rmsclean = (clean ** 2).mean(axis=0) ** 0.5
+        rmsclean = np.where(np.isclose(rmsclean, 0), self._epsilon, rmsclean)
+        scalarclean = 10 ** (norm_to_db / 20) / rmsclean
+        clean = clean * scalarclean
+        rmsclean = (clean ** 2).mean(axis=0) ** 0.5
+
+        rmsnoise = (noise ** 2).mean(axis=0) ** 0.5
+        rmsnoise = np.where(np.isclose(rmsnoise, 0), self._epsilon, rmsnoise)
+        scalarnoise = 10 ** (norm_to_db / 20) / rmsnoise
+        noise = noise * scalarnoise
+        rmsnoise = (noise ** 2).mean(axis=0) ** 0.5
+        rmsnoise = np.where(np.isclose(rmsnoise, 0), self._epsilon, rmsnoise)
+
+        # Set the noise level for a given SNR
+        noisescalar = np.sqrt(rmsclean / (10 ** (snr / 20)) / rmsnoise)
+        noisenewlevel = noise * noisescalar
+        noisyspeech = clean + noisenewlevel
+
+        return clean, noisenewlevel, noisyspeech
+
+    def norm_audio_to_db(self, x, norm_to_db):
+        rms = (x ** 2).mean(axis=0) ** 0.5
+        rms = np.where(np.isclose(rms, 0), self._epsilon, rms)
+        scalar = 10 ** (norm_to_db / 20.0) / rms
+        return x * scalar
+
+    def concatenate_noise_sample(self, clean, noise, fs, silence_length=0.25):
+        while len(noise) < len(clean):
+            if noise.ndim > 1:
+                zeros = np.zeros((int(fs * silence_length), noise.shape[-1]))
+            else:
+                zeros = np.zeros((int(fs * silence_length),))
+            noiseconcat = np.append(noise, zeros, axis=0)
+            noise = np.append(noiseconcat, noise, axis=0)
+
+        return noise
+
+    def perturb_with_input_noise(self, data, noise, data_rms=None, ref_mic=0, norm_to_db=-25.0):
+        """
+        Args:
+            data (AudioSegment): audio data
+            noise (AudioSegment): noise data
+            data_rms (Union[float, List[float]): rms_db for data input
+            ref_mic (int): reference mic index for scaling multi-channel audios
+            norm_to_db (float): will normalise all audio to this DB
+        """
+        if data.num_channels != noise.num_channels:
+            raise ValueError(
+                f"Found mismatched channels for data ({data.num_channels}) and noise ({noise.num_channels})."
+            )
+
+        if not (0 <= ref_mic < data.num_channels):
+            raise ValueError(
+                f" reference mic ID must be an integer in [0, {data.num_channels}), got {ref_mic} instead."
+            )
+
+        if self._snr_samples:
+            snr_db = random.sample(self._snr_samples, 1)[0]
+        else:
+            snr_db = random.uniform(self._min_snr_db, self._max_snr_db)
+        if data_rms is None:
+            data_rms = data.rms_db
+
+        if norm_to_db is None:
+            norm_to_db = data_rms
```

Comment:
It is a vector, but `norm_audio_to_db` works fine if you pass a vector for the db value instead of a scalar, I just checked it. It then applies the corresponding value of the vector to each channel

Reply:
Okay, I understand now, I have revised the code to implement your approach: if ref_mic=None, it will use the entire vector, otherwise it will index into the channel specified by ref_mic